### PR TITLE
Canvas: Add snapping to vertex edit

### LINKED ISSUE #83269: Canvas: Angle snapping when editing mid point
Provides initial support for orthogonal arrows.

1. Snap to vertical / horizontal lines
2. Vertex removal for straight line (snaps as well)

### PR DESCRIPTION
**Pull request by drew08t**
Add snapping to [vertex edit of connections](https://github.com/grafana/grafana/pull/83653) in canvas.

- [x] Horizontal snapping during edit
- [x] Vertical snapping during edit
- [x] Snapping to delete vertex (straight line)
- [x] horizontal snapping during creation
- [x] vertical snapping during creation
- [x] escape snapping hot key (control key)

Disable snapping using control key:
Mar-15-2024 15-20-04

Snapping during vertex creation:
Mar-15-2024 15-19-31

Horizontal and vertical snap during vertex edit:
Mar-13-2024 22-14-16

Deleting a vertex:
Mar-13-2024 15-16-06

Closes #83269

### CODE REVIEW

File: public/app/plugins/panel/canvas/components/connections/Connections.tsx [hunk 1]
```diff
@@ -323,9 +331,52 @@ export class Connections {
...
+      }
+
+      const angleOverall = calculateAngle(vx1, vy1, vx2, vy2);
+      const angleBefore = calculateAngle(vx1, vy1, x, y);
+      deleteVertex = Math.abs(angleBefore - angleOverall) < CONNECTION_VERTEX_SNAP_TOLERANCE;
+    }
+
+    if (deleteVertex) {
+      // Display temporary vertex removal
+      this.scene.selecto!.rootContainer!.style.cursor = 'not-allowed';
```
Comment (nmarrs) [inline public/app/plugins/panel/canvas/components/connections/Connections.tsx:372]: is the cursor style change there to indicate that the vertexes are being removed? I'm not sure this is necessary / might be confusing to end user as it seems like they are doing an action that isn't allowed or something 🤔 

https://github.com/grafana/grafana/assets/22381771/7ef8216d-b7b8-484d-8ade-e69b34fc9fcb

**Review by nmarrs — APPROVED** [review]
Overall LGTM! Great job on the vertex work!! It is super fun to play around with :)

**Comment by lukasztyrala** [comment]
LGTM, great that we have snapping. 👍

File: public/app/plugins/panel/canvas/components/connections/Connections.tsx [hunk 2]
```diff
@@ -18,6 +19,11 @@ import {
 import { CONNECTION_ANCHOR_ALT, ConnectionAnchors, CONNECTION_ANCHOR_HIGHLIGHT_OFFSET } from './ConnectionAnchors';
 import { ConnectionSVG } from './ConnectionSVG';
 
+export const CONNECTION_VERTEX_ID = 'vertex';
+export const CONNECTION_VERTEX_ADD_ID = 'vertexAdd';
+const CONNECTION_VERTEX_ORTHO_TOLERANCE = 0.05; // Cartesian ratio against vertical or horizontal tolerance
```
Comment (codeincarnate) [inline public/app/plugins/panel/canvas/components/connections/Connections.tsx:24]: Lovely use of constants ❤️

File: public/app/plugins/panel/canvas/components/connections/Connections.tsx (hunk 1 above)
Comment (codeincarnate) [reply public/app/plugins/panel/canvas/components/connections/Connections.tsx:372]: I want to agree here, feels like this cursor doesn't quite match up with the outcome expected. Maybe would be better to have a treatment for the line or something like that?

**Review by codeincarnate — APPROVED** [review]
Looks pretty good! See my comment regaring deletion :smile:

File: public/app/plugins/panel/canvas/components/connections/Connections.tsx (hunk 1 above)
Comment (drew08t) [reply public/app/plugins/panel/canvas/components/connections/Connections.tsx:372]: Agreed, appreciate having more eyes on this!

Comment (drew08t) [reply public/app/plugins/panel/canvas/components/connections/Connections.tsx]: It's subtle, but by simply hiding the vertex during "deletion snap", I think is good enough for now.

Mar-20-2024 23-29-20

Comment (nmarrs) [reply public/app/plugins/panel/canvas/components/connections/Connections.tsx]: :shipit:

**Pull request merged**